url29ff2e53f3_nicolas3470_Settlers_of_Catan_AI_tgz-pJ8-robotGameJ8/src/soc/robot/SOCRobotDM.java

```diff
@@ -0,0 +1,2114 @@
+    SOCPlayerTracker.updateWinGameETAs(trackersCopy);
+    float score = calcWGETABonus(playerTrackers, trackersCopy);
+
+    if (!posRoad.getThreats().isEmpty()) {
+      score *= threatMultiplier;
+      D.ebugPrintln("***  (THREAT MULTIPLIER) score * "+threatMultiplier+" = "+score);
+    }
+    D.ebugPrintln("*** ETA for road = "+roadETA);
+    float etaBonus = getETABonus(roadETA, leadersCurrentWGETA, score);
+    D.ebugPrintln("$$$ score = "+score);
+    D.ebugPrintln("etaBonus = "+etaBonus);
+    posRoad.addToScore(etaBonus);
+
+    if ((brain != null) && (brain.getDRecorder().isOn())) {
+      brain.getDRecorder().record("ETA = "+roadETA);
+      brain.getDRecorder().record("WGETA Score = "+df1.format(score));
+      brain.getDRecorder().record("Total road score = "+df1.format(etaBonus));
+    } 
+    
+    D.ebugPrintln("--- after [end] ---");
+    SOCPlayerTracker.undoTryPutPiece(tmpRoad1, game);
+    ourPlayerData.getResources().clear();
+    ourPlayerData.getResources().add(originalResources);
+    D.ebugPrintln("--- cleanup done ---");
+		
+    return etaBonus;
+  }
+
+  /**
+   * calc the win game eta bonus
+   *
+   * @param  trackersBefore   list of player trackers before move
+   * @param  trackersAfter    list of player trackers after move
+   */
+  protected float calcWGETABonus(HashMap trackersBefore, HashMap trackersAfter) {
+    D.ebugPrintln("^^^^^ calcWGETABonus");
+    int originalWGETAs[] = new int[game.maxPlayers];	 
+    int WGETAdiffs[] = new int[game.maxPlayers];	 
+    Vector leaders = new Vector();
+    int bestWGETA = 1000;
+    float bonus = 0;
+
+    Iterator trackersBeforeIter = trackersBefore.values().iterator();
+    while (trackersBeforeIter.hasNext()) {
+      SOCPlayerTracker trackerBefore = (SOCPlayerTracker)trackersBeforeIter.next();
+      D.ebugPrintln("$$$ win game ETA for player "+trackerBefore.getPlayer().getPlayerNumber()+" = "+trackerBefore.getWinGameETA());
+      originalWGETAs[trackerBefore.getPlayer().getPlayerNumber()] = trackerBefore.getWinGameETA();
+      WGETAdiffs[trackerBefore.getPlayer().getPlayerNumber()] = trackerBefore.getWinGameETA();
+
+      if (trackerBefore.getWinGameETA() < bestWGETA) {
+	bestWGETA = trackerBefore.getWinGameETA();
+	leaders.removeAllElements();
+	leaders.addElement(trackerBefore);
+      } else if (trackerBefore.getWinGameETA() == bestWGETA) {
+	leaders.addElement(trackerBefore);
+      }
+    }
+		
+    D.ebugPrintln("^^^^ bestWGETA = "+bestWGETA);
+
+    bonus = calcWGETABonusAux(originalWGETAs, trackersAfter, leaders);
+
+    D.ebugPrintln("^^^^ final bonus = "+bonus);
+
+    return bonus;
+  }
+
+  /**
+   * calcWGETABonusAux
+   *
+   * @param originalWGETAs   the original WGETAs
+   * @param trackersAfter    the playerTrackers after the change
+   * @param leaders          a list of leaders
+   */
+  public float calcWGETABonusAux(int[] originalWGETAs, HashMap trackersAfter, 
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 38. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 1860-1942 into a separate method.